File: inkcluded-405/DrawView.m
//
//  DrawView.m
//  inkluded-405
//
//  Created by Josh Choi on 1/22/17.
//  Copyright © 2017 Josh Choi. All rights reserved.
//

#import "DrawView.h"
#import "Stroke.h"
#import <WILLCore/WILLCore.h>

@implementation DrawView
{
    // Renderer of layers
    WCMRenderingContext * willContext;
    
    // The main viewable layer
    WCMLayer* viewLayer;
    
    // Strokes layer
    WCMLayer * strokesLayer;
    
    // Renderer of all the layers
    WCMStrokeRenderer * strokeRenderer;
    
    // Helps build the path of strokes
    WCMSpeedPathBuilder * pathBuilder;
    
    // Help smooth out the paths
    WCMMultiChannelSmoothener * pathSmoothener;
    
    // Path brush for redrawing strokes
    WCMStrokeBrush * pathBrush;
    
    // Array of all the strokes
    NSMutableArray * strokes;
    
    // Stride of the brush, for redrawing
    int pathStride;
}

// The layer class that is being wrapped by the WILL framework
+ (Class) layerClass
{
    return [CAEAGLLayer class];
}

/**
 * Constructor that creates a view based on the a given size
 *
 * @param frame the size to initialize the view to
 * @return the DrawView object
 **/
-(id) initWithFrame:(CGRect)frame
{
    strokes = [NSMutableArray array];
    self = [super initWithFrame:frame];
    // checks if the UIView was successfully created
    if (self)
    {
        [self initWillContext];
        [willContext setTarget:strokesLayer];
        [willContext clearColor:[UIColor clearColor]];
        
        [self refreshViewInRect:viewLayer.bounds];
    }
    return self;
}

/**
 * Initializes the DrawView resources with default initial values.  Initialize the resources for the DrawView to
 * successfully start drawing.
 *
 **/
- (void) initWillContext
{
    // checks if the renderer was not initialized yet
    if (!willContext)
    {
        self.contentScaleFactor = [UIScreen mainScreen].scale;
        
        CAEAGLLayer *eaglLayer = (CAEAGLLayer *)self.layer;
        eaglLayer.opaque = YES;
        eaglLayer.drawableProperties = [NSDictionary dictionaryWithObjectsAndKeys:
                                        [NSNumber numberWithBool:YES], kEAGLDrawablePropertyRetainedBacking,
                                        kEAGLColorFormatRGBA8, kEAGLDrawablePropertyColorFormat, nil];
        
        EAGLContext* eaglContext = [[EAGLContext alloc] initWithAPI:kEAGLRenderingAPIOpenGLES2];
        
        // Checks to make sure the EAGLContext was correctly constructed
        if (!eaglContext || ![EAGLContext setCurrentContext:eaglContext])
        {
            NSLog(@"Unable to create EAGLContext!");
            return;
        }
        
        willContext = [WCMRenderingContext contextWithEAGLContext:eaglContext];
        
        viewLayer = [willContext layerFromEAGLDrawable:(id<EAGLDrawable>)self.layer withScaleFactor:self.contentScaleFactor];
        
        strokesLayer = [willContext layerWithWidth:viewLayer.bounds.size.width andHeight:viewLayer.bounds.size.height andScaleFactor:viewLayer.scaleFactor andUseTextureStorage:YES];
        
        pathBrush = [willContext solidColorBrush];
        
        pathBuilder = [[WCMSpeedPathBuilder alloc] init];
        [pathBuilder setNormalizationConfigWithMinValue:0 andMaxValue:7000];
        [pathBuilder setPropertyConfigWithName:WCMPropertyNameWidth andMinValue:2 andMaxValue:15 andInitialValue:NAN andFinalValue:NAN andFunction:WCMPropertyFunctionPower andParameter:1 andFlip:NO];
        
        pathStride = [pathBuilder calculateStride];
        
        pathSmoothener = [[WCMMultiChannelSmoothener alloc] initWithChannelsCount:pathStride];
        
        strokeRenderer =  [willContext  strokeRendererWithSize:viewLayer.bounds.size andScaleFactor:viewLayer.scaleFactor];
        
        strokeRenderer.brush = [willContext solidColorBrush];
        strokeRenderer.stride = pathStride;
        strokeRenderer.color = [UIColor blackColor];
    }
}

/**
 * Refreshes the DrawView to correctly show the layers. Updates
 * the viewLayer.
 *
 * @param rect The area to update
 **/
-(void) refreshViewInRect:(CGRect)rect
{
    [willContext setTarget:viewLayer andClipRect:rect];
    [willContext clearColor:[UIColor whiteColor]];
    
    [willContext drawLayer:strokesLayer withSourceRect:rect andDestinationRect:rect andBlendMode:WCMBlendModeNormal];
    [strokeRenderer blendStrokeUpdatedAreaInLayer:viewLayer withBlendMode:WCMBlendModeNormal];
    
    [viewLayer present];
}

/**
 * Processes touches on the DrawView.  Constantly called whenever the DrawView is touched.
 * Creates strokes and and the end of a series touches, saves the touches as a single stroke.
 *
 * @param touches a representation of the touch
 * @param event unused so far
 **/
- (void) processTouches:(NSSet *)touches withEvent:(UIEvent *)event
{
    UITouch * touch = [touches anyObject];
    
    // Check if the touches are not moving
    if (touch.phase != UITouchPhaseStationary)
    {
        CGPoint location = [touch locationInView:self];
        
        WCMInputPhase wcmInputPhase;
        
        // If the touches have just started
        if (touch.phase == UITouchPhaseBegan)
        {
            wcmInputPhase = WCMInputPhaseBegin;
            
            [pathSmoothener reset];
        }
        // If the touches have begun moving again
        else if (touch.phase == UITouchPhaseMoved)
        {
            wcmInputPhase = WCMInputPhaseMove;
        }
        // The touches have stopped, pen is off the canvas.
        else if (touch.phase == UITouchPhaseEnded || touch.phase == UITouchPhaseCancelled)
        {
            wcmInputPhase = WCMInputPhaseEnd;
        }
        
        WCMFloatVectorPointer * points = [pathBuilder addPointWithPhase:wcmInputPhase andX:location.x andY:location.y andTimestamp:touch.timestamp];
        WCMFloatVectorPointer * smoothedPoints = [pathSmoothener smoothValues:points reachFinalValues:wcmInputPhase == WCMInputPhaseEnd];
        WCMPathAppendResult* pathAppendResult = [pathBuilder addPathPart:smoothedPoints];
        
        WCMFloatVectorPointer * prelimPoints = [pathBuilder createPreliminaryPath];
        WCMFloatVectorPointer * smoothedPrelimPoints = [pathSmoothener smoothValues:prelimPoints reachFinalValues:YES];
        WCMFloatVectorPointer * prelimPath = [pathBuilder finishPreliminaryPath:smoothedPrelimPoints];
        
        [strokeRenderer drawPoints:pathAppendResult.addedPath finishStroke:wcmInputPhase == WCMInputPhaseEnd];
        [strokeRenderer drawPreliminaryPoints:prelimPath];
        
        [self refreshViewInRect:strokeRenderer.updatedArea];
        
        // Check to see if all inputs from a single stroke have stopped to store that single stroke, pen is off canvas.
        if (wcmInputPhase == WCMInputPhaseEnd)
        {
            [strokeRenderer blendStrokeInLayer:strokesLayer withBlendMode:WCMBlendModeNormal];
            
            Stroke* stroke = [Stroke strokeWithPoints:[WCMFloatVector vectorWithBegin:pathAppendResult.wholePath.begin andEnd:pathAppendResult.wholePath.end]
                                            andStride:pathStride
                                             andWidth:NAN
                                             andColor:[UIColor blackColor]
                                                andTs:0
                                                andTf:1
                                         andBlendMode:WCMBlendModeNormal];
            [strokes addObject:stroke];
        }
    }
    
    
}

/**
 * Saves strokes of this current DrawView.  Currently creates the will doc, may need to be
 * changed in the future once other layers are added so only 1 doc is created and that
 * one doc is consistently updated. The default will doc path is:
 * /Users/<USER>/Library/Developer/CoreSimulator/Devices/<SIMULATOR-ID>/data/Containers/Data/Application/<APP-ID>/Documents/
 *
 **/
- (void) saveStrokes
{
    WCMInkEncoder * inkEncoder = [[WCMInkEncoder alloc] init];
    
    // For every stroke in strokes, we will encode that strokes attributes.
    for (Stroke * s in strokes)
    {
        [inkEncoder encodePathWithPrecision:2 andPoints:s.points andStride:s.stride andWidth:s.width andColor:s.color andTs:0 andTf:1 andBlendMode:WCMBlendModeNormal];
    }
    
    NSData * inkData = [inkEncoder getBytes];
    
    WCMDocument * doc = [[WCMDocument alloc] init];
    
    WCMDocumentSection * section = [[WCMDocumentSection alloc] init];
    section.size = self.bounds.size;
    
    WCMDocumentSectionPaths * pathsElement = [[WCMDocumentSectionPaths alloc] init];
    [pathsElement.content setData:inkData withType:[WCMDocumentContentType STROKES]];
    
    [doc.sections addObject:section];
    [section addElement:pathsElement];
    
    NSString *documentsPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) objectAtIndex:0];
    NSString *willDocPath = [documentsPath stringByAppendingPathComponent:@"document.will"];
    
    [doc createDocumentAtPath:willDocPath];
}

/**
 * Decode all the strokes from the default will doc path.  May also need to be changed to
 * handle extra layers and resources that the will doc will store.  The default will doc
 * path is: 
 * /Users/<USER>/Library/Developer/CoreSimulator/Devices/<SIMULATOR-ID>/data/Containers/Data/Application/<APP-ID>/Documents/
 *
 **/
- (void) decodeStrokesFromDocumentPath
{
    NSString *documentsPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) objectAtIndex:0];
    NSString *path = [documentsPath stringByAppendingPathComponent:@"document.will"];
    
    WCMDocument * doc = [[WCMDocument alloc] init];
    
    [doc loadDocumentAtPath:path];
    
    WCMDocumentSection * section = doc.sections[0];
    WCMDocumentSectionPaths * pathsElement = section.subelements[0];
    NSData * inkData = [pathsElement.content loadData];
    
    WCMInkDecoder * decoder = [[WCMInkDecoder alloc] initWithData:inkData];
    
    WCMFloatVector* strokePoints;
    unsigned int strokeStride;
    float strokeWidth;
    UIColor* strokeColor;
    float strokeStartValue;
    float strokeFinishValue;
    WCMBlendMode blendMode;
    
    strokes = [[NSMutableArray alloc] init];
    
    // While there are still strokes in the will doc to draw, we will continue adding strokes.
    while([decoder decodePathToPoints:&strokePoints
                            andStride:&strokeStride
                             andWidth:&strokeWidth
                             andColor:&strokeColor
                                andTs:&strokeStartValue
                                andTf:&strokeFinishValue
                         andBlendMode:&blendMode])
    {
        Stroke * stroke = [Stroke strokeWithPoints:strokePoints
                                         andStride:strokeStride
                                          andWidth:strokeWidth
                                          andColor:strokeColor
                                             andTs:strokeStartValue
                                             andTf:strokeFinishValue
                                      andBlendMode:blendMode];
        [strokes addObject:stroke];
    }
    
    [willContext setTarget:strokesLayer];
    [willContext clearColor:[UIColor clearColor]];
    
    WCMStrokeRenderer * renderer = [willContext strokeRendererWithSize:viewLayer.bounds.size andScaleFactor:viewLayer.scaleFactor];
    renderer.brush = pathBrush;
    
    // For every stroke in strokes, we will render it to the strokesLayer.
    for (Stroke * s in strokes)
    {
        renderer.stride = s.stride;
        renderer.width = s.width;
        renderer.color = s.color;
        renderer.ts = s.ts;
        renderer.tf = s.tf;
        
        [renderer resetAndClearBuffers];
        [renderer drawPoints:s.points.pointer finishStroke:YES];
        [renderer blendStrokeInLayer:strokesLayer withBlendMode:s.blendMode];
    }
    
    [self refreshViewInRect:viewLayer.bounds];
    
}

/**
 * The next 4 functions are immediately called when the DrawView is touched.
 *
 **/

-(void) touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
{
    [self processTouches:touches withEvent:event];
}

-(void) touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event
{
    [self processTouches:touches withEvent:event];
}

-(void) touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event
{
    [self processTouches:touches withEvent:event];
}

-(void) touchesCancelled:(NSSet *)touches withEvent:(UIEvent *)event
{
    [self processTouches:touches withEvent:event];
}

@end

File: inkcluded-405/DrawView.swift
//
//  DrawView.swift
//  inkcluded-405
//
//  Created by Francis Yuen on 2/22/17.
//  Copyright © 2017 Boba. All rights reserved.
//

import UIKit

class DrawView: UIView {
    
    // Renderer of layers
    private var willContext: WCMRenderingContext!
    
    // The main viewable layer
    private var viewLayer: WCMLayer!
    
    // Strokes layer
    private var strokesLayer: WCMLayer!
    
    // Renderer for stroke layer
    private var strokeRenderer: WCMStrokeRenderer!
    
    // Helps build the path of strokes
    private var pathBuilder: WCMSpeedPathBuilder!
    
    // Help smooth out the paths
    private var pathSmoother: WCMMultiChannelSmoothener!
    
    // Path brush for redrawing strokes
    private var pathBrush: WCMStrokeBrush!
    
    // Delegate for saving strokes data
    private var delegate: DrawStrokesDelegate!
    
    private var pathStride: Int32!
    
    // Whether to allow strokes to be drawn
    var shouldDraw: Bool
    
    override init(frame: CGRect) {
        shouldDraw = true
        super.init(frame: frame)

        self.initWillContext()
        self.backgroundColor = UIColor.white
        willContext.setTarget(strokesLayer)
        willContext.clear(UIColor.clear)
    }
    
    // The layer class that is being wrapped by the WILL framework
    override class var layerClass: AnyClass {
        return CAEAGLLayer.self
    }

    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func initWillContext() {
        if (willContext == nil) {
            self.contentScaleFactor = UIScreen.main.scale
            
            let eaglLayer = self.layer as! CAEAGLLayer
            eaglLayer.isOpaque = true
            eaglLayer.drawableProperties = [
                kEAGLDrawablePropertyRetainedBacking: NSNumber(booleanLiteral: true),
                kEAGLDrawablePropertyColorFormat: kEAGLColorFormatRGBA8
            ]
            
            let eaglContext = EAGLContext(api: .openGLES2)
            
            if (eaglContext == nil || !EAGLContext.setCurrent(eaglContext)) {
                print("Unable to create EAGLContext")
                return
            }
            
            self.willContext = WCMRenderingContext(eaglContext: eaglContext)
            
            self.viewLayer = willContext.layer(
                                from: self.layer as! EAGLDrawable,
                                withScaleFactor: Float(self.contentScaleFactor)
                             )
            
            self.strokesLayer = willContext.layer(
                                    withWidth: Int32(self.viewLayer.bounds.size.width),
                                    andHeight: Int32(self.viewLayer.bounds.size.height),
                                    andScaleFactor: Int32(self.viewLayer.scaleFactor),
                                    andUseTextureStorage: true
                                )
            
            self.pathBrush = willContext.solidColorBrush()
            
            self.pathBuilder = WCMSpeedPathBuilder()
            self.pathBuilder.setNormalizationConfigWithMinValue(0, andMaxValue: 7000)
            self.pathBuilder.setPropertyConfigWith(
                WCMPropertyName.width,
                andMinValue: 2,
                andMaxValue: 15,
                andInitialValue: Float.nan,
                andFinalValue: Float.nan,
                andFunction: WCMPropertyFunction.power,
                andParameter: 1,
                andFlip: false
            )
            
            self.pathStride = pathBuilder.calculateStride()
            
            self.pathSmoother = WCMMultiChannelSmoothener(channelsCount: self.pathStride)
            
            self.strokeRenderer = willContext.strokeRenderer(
                                      with: self.viewLayer.bounds.size,
                                      andScaleFactor: CGFloat(self.viewLayer.scaleFactor)
                                  )
            self.strokeRenderer.brush = willContext.solidColorBrush()
            self.strokeRenderer.stride = pathStride
            self.strokeRenderer.color = UIColor.black
        }
    }
    
    func setNewDelegate(newDelegate: DrawStrokesDelegate) {
        self.delegate = newDelegate
    }
    
    func refreshViewInRect(rect: CGRect) {
        willContext.setTarget(self.viewLayer, andClipRect: rect)
        willContext.clear(UIColor.white)
        
        willContext.draw(
            self.strokesLayer,
            withSourceRect: rect,
            andDestinationRect: rect,
            andBlendMode: WCMBlendMode.normal
        )
        strokeRenderer.blendStrokeUpdatedArea(in: self.viewLayer, with: WCMBlendMode.normal)
        
        viewLayer.present()
    }
    
    func processTouches(touches: Set<UITouch>, withEvent event: UIEvent) {
        let touch = touches[touches.index(touches.startIndex, offsetBy: 0)]
        
        if (shouldDraw && touch.phase != UITouchPhase.stationary) {
            let location = touch.location(in: self)
            var wcmInputPhase: WCMInputPhase? = nil
            
            if (touch.phase == UITouchPhase.began) {
                wcmInputPhase = WCMInputPhase.begin
                pathSmoother.reset()
            }
            else if (touch.phase == UITouchPhase.moved) {
                wcmInputPhase = WCMInputPhase.move
            }
            else if (touch.phase == UITouchPhase.ended || touch.phase == UITouchPhase.cancelled) {
                wcmInputPhase = WCMInputPhase.end
            }
            
            let points = pathBuilder.addPoint(
                            with: wcmInputPhase!,
                            andX: Float(location.x),
                            andY: Float(location.y),
                            andTimestamp: touch.timestamp
                         )
            let smoothedPoints = pathSmoother.smoothValues(
                                    points,
                                    reachFinalValues: wcmInputPhase == WCMInputPhase.end
                                 )
            let pathAppendResult = pathBuilder.addPathPart(smoothedPoints)
            
            let prelimPoints = pathBuilder.createPreliminaryPath()
            let smoothedPrelimPoints = pathSmoother.smoothValues(prelimPoints, reachFinalValues: true)
            let prelimPath = pathBuilder.finishPreliminaryPath(smoothedPrelimPoints)
            
            strokeRenderer.drawPoints(pathAppendResult?.addedPath, finishStroke: wcmInputPhase == WCMInputPhase.end)
            strokeRenderer.drawPreliminaryPoints(prelimPath)
            
            self.refreshViewInRect(rect: strokeRenderer.updatedArea)
            
            if (wcmInputPhase == WCMInputPhase.end) {
                strokeRenderer.blendStroke(in: strokesLayer, with: WCMBlendMode.normal)
                
                let stroke = Stroke(points: WCMFloatVector(
                                        begin: pathAppendResult?.wholePath.begin(),
                                        andEnd: pathAppendResult?.wholePath.end()),
                                    andStride: pathStride,
                                    andWidth: Float.nan,
                                    andColor: UIColor.black,
                                    andTs: 0,
                                    andTf: 1,
                                    andBlendMode: WCMBlendMode.normal)
                delegate.addStroke(stroke: stroke!)
            }
        }
    }
    
    func scaleStrokeVector(vector: WCMFloatVector, atScaleFactor scaleFactor: Float) -> WCMFloatVector {
        var cur = vector.begin()
        
        while (cur != vector.end()) {
            cur?.pointee = (cur?.pointee)! * scaleFactor
            cur = cur?.advanced(by: 1)
        }
    
        return vector
    }
    
    func determineScaleFactor(originalSize size: CGSize) -> Float {
        var newSize = CGSize(width: size.width, height: size.height)
        if (size.width == 0 || size.height == 0) {
            // Hacky solution to resize drawings from the windows team bc their WILL SDK doesn't save size
            newSize.width = 650
            newSize.height = 650
        }
    
        return Float(self.frame.size.width / newSize.width)
    }
    
    func refreshViewWithElements(elements: [AnyObject], atSize size: CGSize) -> [AnyObject] {
        var scaledElements: [AnyObject] = []
        
        for element in elements {
            let scaleFactor = determineScaleFactor(originalSize: size)
            if let strokeElement = element as? Stroke {
                let scaledVector = scaleStrokeVector(
                    vector: strokeElement.points.copy(),
                    atScaleFactor: scaleFactor
                )
                strokeRenderer.resetAndClearBuffers()
                strokeRenderer.drawPoints(scaledVector.pointer(), finishStroke: true)
                strokeRenderer.blendStroke(in: strokesLayer, with: strokeElement.blendMode)
                scaledElements.append(Stroke(points: scaledVector, andStride: strokeElement.stride, andWidth: strokeElement.width, andColor: strokeElement.color, andTs: strokeElement.ts, andTf: strokeElement.tf, andBlendMode: strokeElement.blendMode))
            }
            else if let messageImageElement = element as? DraggableImageView {
                let imageElement = DraggableImageView(image: messageImageElement.image!)
                imageElement.isUserInteractionEnabled = shouldDraw
                imageElement.frame = CGRect(origin: CGPoint(x: messageImageElement.frame.origin.x * CGFloat(scaleFactor),
                                                            y: messageImageElement.frame.origin.y * CGFloat(scaleFactor)),
                                            size: CGSize(width: messageImageElement.frame.width * CGFloat(scaleFactor),
                                                         height: messageImageElement.frame.height * CGFloat(scaleFactor)))
                scaledElements.append(imageElement)
                self.addSubview(imageElement)
            }
            else if let messageTextElement = element as? DraggableTextView {
                let textViewElement = DraggableTextView(frame: CGRect(origin: CGPoint(x: messageTextElement.frame.origin.x * CGFloat(scaleFactor),
                                                                                      y: messageTextElement.frame.origin.y * CGFloat(scaleFactor)),
                                                                      size: CGSize(width: messageTextElement.frame.width * CGFloat(scaleFactor),
                                                                                   height: messageTextElement.frame.height * CGFloat(scaleFactor))))
                textViewElement.text = messageTextElement.text
                textViewElement.isUserInteractionEnabled = shouldDraw
                scaledElements.append(textViewElement)
                self.addSubview(textViewElement)
            }
            else {
                print("Not expecting this type")
            }
        }
        
        refreshViewInRect(rect: viewLayer.bounds)
        return scaledElements
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.processTouches(touches: touches, withEvent: event!)
        self.endEditing(true)
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.processTouches(touches: touches, withEvent: event!)
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.processTouches(touches: touches, withEvent: event!)
    }
    
    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.processTouches(touches: touches, withEvent: event!)
    }
}
